source/Coop/Mod/GameEnvironmentServer.cs

```diff
@@ -1,38 +1,45 @@
 ﻿using System;
+using System.Collections.Generic;
 using System.Linq;
 using Coop.Mod.Patch;
+using Coop.Mod.Patch.MobilePartyPatches;
 using Coop.Mod.Persistence;
 using Coop.Mod.Persistence.Party;
-using Coop.Mod.Persistence.RPC;
+using Coop.Mod.Persistence.RemoteAction;
 using NLog;
+using RemoteAction;
 using Sync;
 using Sync.Store;
 using TaleWorlds.CampaignSystem;
+using TaleWorlds.ObjectSystem;
 
 namespace Coop.Mod
 {
     public class GameEnvironmentServer : IEnvironmentServer
     {
         private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
 
-        public FieldAccessGroup<MobileParty, MovementData> TargetPosition =>
-            CampaignMapMovement.Movement;
-
-        public bool CanChangeTimeControlMode => CoopServer.Instance.AreAllClientsPlaying;
-
         public EventBroadcastingQueue EventQueue => CoopServer.Instance.Persistence?.EventQueue;
 
-        public MobileParty GetMobilePartyByIndex(int iPartyIndex)
+        private Dictionary<MBGUID, MobileParty> m_PartyCache = new Dictionary<MBGUID, MobileParty>();
+
+        public MobileParty GetMobilePartyById(MBGUID guid)
         {
-            MobileParty ret = null;
-            GameLoopRunner.RunOnMainThread(
-                () =>
-                {
-                    ret = MobileParty.All.SingleOrDefault(p => p.Party.Index == iPartyIndex);
-                });
+            if (!m_PartyCache.TryGetValue(guid, out MobileParty ret))
+            {
+                GameLoopRunner.RunOnMainThread(
+                    () =>
+                    {
+                        // Update the whole cache since we're already in the game loop thread. Doesn't happen that often.
+                        m_PartyCache = MobileParty.All.ToDictionary(party => party.Id);
```

Comment:
LINQ as parallel single would work faster afaik. Might not be due to how the threading is handled

Reply:
0c1de41